test/functional/p2p_handshake.py

```diff
@@ -47,6 +47,7 @@ def test_desirable_service_flags(self, node, service_flag_tests, desirable_servi
            service flags in the VERSION message. The test is exercised for all relevant
            outbound connection types where the desirable service flags check is done."""
         CONNECTION_TYPES = ["outbound-full-relay", "block-relay-only", "addr-fetch"]
+        peer_id = 0
```

Comment:
I think it would be more stable to have a global counter, so that `peer_id` is never reused. Currently, it is still reused over different `test_desirable_service_flags` calls.
E.g., if I add the net sleep and also change `CONNECTION_TYPES = ["outbound-full-relay"]` so that  `test_desirable_service_flags` only makes one connection, I run into the timeout even with your fix.

Reply:
makes sense, less diff too. but `peer_id` is bounded to [0, 11]/`p2p_idx = peer_id + 1` is bounded to [1, 12] [here](https://github.com/bitcoin/bitcoin/blob/dbd2000b34903b87ae2e02eb2fc6c4a2f7d11451/test/functional/test_framework/p2p.py#L745). so we'd cross that limit if we do global counters.

Reply:
oh - I didn't know about that, interesting!
Another thing we could do is to not only do `peer.wait_for_disconnect()` but also wait until the bitcoind side has disconnected - something like `self.wait_until(lambda: len(self.nodes[0].getpeerinfo()) == X)`. Haven't tried that though.

Reply:
>  Another thing we could do is to not only do peer.wait_for_disconnect() but also wait until the bitcoind side has disconnected - something like self.wait_until(lambda: len(self.nodes[0].getpeerinfo()) == X). Haven't tried that though.

yeah this sounds like a better option! updated the PR to use this approach.

>  (that is for the duration of all tests, not the test itself) to avoid port collisions.

oh makes sense, a TIL!